utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2202 @@
+      gcstats->num_deletes++;
+      gcstats->deleted_size += record.GetBlobSize();
+      if (first_gc) continue;
+
+      Transaction* txn = static_cast<OptimisticTransactionDB*>(opt_db_.get())
+                             ->BeginTransaction(write_options_);
+      txn->Delete(cfh, record.Key());
+      Status s1 = txn->Commit();
+      // chances that this DELETE will fail is low. If it fails, it would be
+      // because
+      // a new version of the key came in at this time, which will override
+      // the current version being iterated on.
+      if (s1.IsBusy()) {
+        Log(InfoLogLevel::INFO_LEVEL, db_options_.info_log,
+            "Optimistic transaction failed delete: %s bn: %" PRIu32,
+            bfptr->PathName().c_str(), gcstats->blob_count);
+      } else {
+        Log(InfoLogLevel::DEBUG_LEVEL, db_options_.info_log,
+            "Successfully added delete back into LSM: %s bn: %" PRIu32,
+            bfptr->PathName().c_str(), gcstats->blob_count);
+
+        // assume that failures happen due to new writes.
+        gcstats->succ_deletes_lsm++;
+      }
+      delete txn;
+      continue;
+    } else if (first_gc) {
+      continue;
+    }
+
+    if (!newfile) {
+      // new file
+      std::string reason("GC of ");
+      reason += bfptr->PathName();
+      newfile = NewBlobFile(reason);
+      gcstats->newfile = newfile;
+
+      new_writer = CheckOrCreateWriterLocked(newfile);
+      newfile->header_ = std::move(header);
+      // Can't use header beyond this point
+      newfile->header_valid_ = true;
+      newfile->file_size_ = BlobLogHeader::kHeaderSize;
+      s = new_writer->WriteHeader(newfile->header_);
+
+      if (!s.ok()) {
+        Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+            "File: %s - header writing failed", newfile->PathName().c_str());
+        return s;
+      }
+
+      WriteLock wl(&mutex_);
+
+      dir_change_.store(true);
+      blob_files_.insert(std::make_pair(newfile->BlobFileNumber(), newfile));
+    }
+
+    gcstats->num_relocs++;
+    std::string index_entry;
+
+    uint64_t blob_offset = 0;
+    uint64_t key_offset = 0;
+    // write the blob to the blob log.
+    s = new_writer->AddRecord(record.Key(), record.Blob(), &key_offset,
+                              &blob_offset, record.GetTTL());
+
+    BlobHandle handle;
+    handle.set_filenumber(newfile->BlobFileNumber());
+    handle.set_size(record.Blob().size());
+    handle.set_offset(blob_offset);
+    handle.set_compression(bdb_options_.compression);
+    handle.EncodeTo(&index_entry);
+
+    new_writer->AddRecordFooter(record.GetSN());
+    newfile->blob_count_++;
+    newfile->file_size_ += BlobLogRecord::kHeaderSize + record.Key().size() +
+                           record.Blob().size() + BlobLogRecord::kFooterSize;
+
+    Transaction* txn = static_cast<OptimisticTransactionDB*>(opt_db_.get())
+                           ->BeginTransaction(write_options_);
+    txn->Put(cfh, record.Key(), index_entry);
```

Comment:
What if Put() happened before the transaction starts?
Should we start the transaction (or keep the previous one) before line 1771?